// @datadog/datadog-ci/dist/commands/synthetics/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_COMMAND_CONFIG = exports.utils = exports.execute = exports.executeTests = exports.JUnitReporter = exports.DefaultReporter = exports.CriticalError = exports.CiError = void 0;
const run_tests_lib_1 = require("./run-tests-lib");
var errors_1 = require("./errors");
Object.defineProperty(exports, "CiError", { enumerable: true, get: function () { return errors_1.CiError; } });
Object.defineProperty(exports, "CriticalError", { enumerable: true, get: function () { return errors_1.CriticalError; } });
__exportStar(require("./interfaces"), exports);
var default_1 = require("./reporters/default");
Object.defineProperty(exports, "DefaultReporter", { enumerable: true, get: function () { return default_1.DefaultReporter; } });
var junit_1 = require("./reporters/junit");
Object.defineProperty(exports, "JUnitReporter", { enumerable: true, get: function () { return junit_1.JUnitReporter; } });
var run_tests_lib_2 = require("./run-tests-lib");
Object.defineProperty(exports, "executeTests", { enumerable: true, get: function () { return run_tests_lib_2.executeTests; } });
Object.defineProperty(exports, "execute", { enumerable: true, get: function () { return run_tests_lib_2.execute; } });
exports.utils = __importStar(require("./utils/public"));
exports.DEFAULT_COMMAND_CONFIG = (0, run_tests_lib_1.getDefaultConfig)();
