// @datadog/datadog-ci/dist/commands/synthetics/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecutionRule = exports.Operator = void 0;
var Operator;
(function (Operator) {
    Operator["contains"] = "contains";
    Operator["doesNotContain"] = "doesNotContain";
    Operator["is"] = "is";
    Operator["isNot"] = "isNot";
    Operator["isInLessThan"] = "isInLessThan";
    Operator["isInMoreThan"] = "isInMoreThan";
    Operator["lessThan"] = "lessThan";
    Operator["lessThanOrEqual"] = "lessThanOrEqual";
    Operator["moreThan"] = "moreThan";
    Operator["moreThanOrEqual"] = "moreThanOrEqual";
    Operator["matches"] = "matches";
    Operator["doesNotMatch"] = "doesNotMatch";
    Operator["validatesJSONPath"] = "validatesJSONPath";
    Operator["validatesXPath"] = "validatesXPath";
})(Operator || (exports.Operator = Operator = {}));
var ExecutionRule;
(function (ExecutionRule) {
    ExecutionRule["BLOCKING"] = "blocking";
    ExecutionRule["NON_BLOCKING"] = "non_blocking";
    ExecutionRule["SKIPPED"] = "skipped";
})(ExecutionRule || (exports.ExecutionRule = ExecutionRule = {}));
