// @google-cloud/run/build/src/v2/index.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerPoolsClient = exports.TasksClient = exports.ServicesClient = exports.RevisionsClient = exports.JobsClient = exports.ExecutionsClient = exports.BuildsClient = void 0;
var builds_client_1 = require("./builds_client");
Object.defineProperty(exports, "BuildsClient", { enumerable: true, get: function () { return builds_client_1.BuildsClient; } });
var executions_client_1 = require("./executions_client");
Object.defineProperty(exports, "ExecutionsClient", { enumerable: true, get: function () { return executions_client_1.ExecutionsClient; } });
var jobs_client_1 = require("./jobs_client");
Object.defineProperty(exports, "JobsClient", { enumerable: true, get: function () { return jobs_client_1.JobsClient; } });
var revisions_client_1 = require("./revisions_client");
Object.defineProperty(exports, "RevisionsClient", { enumerable: true, get: function () { return revisions_client_1.RevisionsClient; } });
var services_client_1 = require("./services_client");
Object.defineProperty(exports, "ServicesClient", { enumerable: true, get: function () { return services_client_1.ServicesClient; } });
var tasks_client_1 = require("./tasks_client");
Object.defineProperty(exports, "TasksClient", { enumerable: true, get: function () { return tasks_client_1.TasksClient; } });
var worker_pools_client_1 = require("./worker_pools_client");
Object.defineProperty(exports, "WorkerPoolsClient", { enumerable: true, get: function () { return worker_pools_client_1.WorkerPoolsClient; } });
