// @datadog/datadog-ci/dist/commands/sarif/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadSarifReportCommand = void 0;
const fs_1 = __importDefault(require("fs"));
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const upath_1 = __importDefault(require("upath"));
const constants_1 = require("../../constants");
const concurrency_1 = require("../../helpers/concurrency");
const env_1 = require("../../helpers/env");
const fips_1 = require("../../helpers/fips");
const glob_1 = require("../../helpers/glob");
const retry_1 = require("../../helpers/retry");
const tags_1 = require("../../helpers/tags");
const utils_1 = require("../../helpers/utils");
const validation = __importStar(require("../../helpers/validation"));
const api_1 = require("./api");
const renderer_1 = require("./renderer");
const utils_2 = require("./utils");
const validation_1 = require("./validation");
class UploadSarifReportCommand extends clipanion_1.Command {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.basePaths = clipanion_1.Option.Rest({ required: 1 });
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.env = clipanion_1.Option.String('--env', 'ci');
        this.maxConcurrency = clipanion_1.Option.String('--max-concurrency', '20', { validator: validation.isInteger() });
        this.serviceFromCli = clipanion_1.Option.String('--service');
        this.tags = clipanion_1.Option.Array('--tags');
        this.gitPath = clipanion_1.Option.String('--git-repository');
        this.noVerify = clipanion_1.Option.Boolean('--no-verify', false);
        this.noCiTags = clipanion_1.Option.Boolean('--no-ci-tags', false);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            env: process.env.DD_ENV,
            envVarTags: process.env.DD_TAGS,
        };
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
            // TODO(julien): remove this notice in April 2025
            if (this.serviceFromCli) {
                this.context.stderr.write('The CLI flag `--service` is deprecated and will be removed in a future version of datadog-ci\n');
                this.context.stderr.write('To associate findings with services, consider using the service-to-repo mapping from service catalog\n');
                this.context.stderr.write('Learn more at https://docs.datadoghq.com/getting_started/code_security/?tab=staticcodeanalysissast#link-datadog-services-to-repository-scan-results\n');
            }
            if (!this.basePaths || !this.basePaths.length) {
                this.context.stderr.write('Missing basePath\n');
                return 1;
            }
            if (!this.config.env) {
                this.config.env = this.env;
            }
            const api = this.getApiHelper();
            // Normalizing the basePath to resolve .. and .
            this.basePaths = this.basePaths.map((basePath) => upath_1.default.normalize(basePath));
            const spanTags = yield (0, tags_1.getSpanTags)(this.config, this.tags, !this.noCiTags, this.gitPath);
            // Gather any missing mandatory git fields to display to the user
            const missingGitFields = (0, tags_1.getMissingRequiredGitTags)(spanTags);
            if (missingGitFields.length > 0) {
                this.context.stdout.write((0, renderer_1.renderMissingTags)(missingGitFields));
                return 1;
            }
            const payloads = yield this.getMatchingSarifReports(spanTags);
            if (payloads.length === 0) {
                this.context.stdout.write((0, renderer_1.renderFilesNotFound)(this.basePaths));
                return 1;
            }
            const sha = spanTags[tags_1.GIT_SHA] || 'sha-not-found';
            const env = this.config.env || 'env-not-set';
            this.context.stdout.write((0, renderer_1.renderCommandInfo)(this.basePaths, env, sha, this.maxConcurrency, this.dryRun, this.noVerify));
            const upload = (payload) => this.uploadSarifReport(api, payload);
            const initialTime = new Date().getTime();
            yield (0, concurrency_1.doWithMaxConcurrency)(this.maxConcurrency, payloads, upload);
            const totalTimeSeconds = (Date.now() - initialTime) / 1000;
            this.context.stdout.write((0, renderer_1.renderSuccessfulCommand)(payloads.length, totalTimeSeconds));
        });
    }
    uploadSarifReport(api, sarifReport) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.dryRun) {
                this.context.stdout.write((0, renderer_1.renderDryRunUpload)(sarifReport));
                return;
            }
            try {
                yield (0, retry_1.retryRequest)(() => api.uploadSarifReport(sarifReport, this.context.stdout.write.bind(this.context.stdout)), {
                    onRetry: (e, attempt) => {
                        this.context.stderr.write((0, renderer_1.renderRetriedUpload)(sarifReport, e.message, attempt));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.context.stderr.write((0, renderer_1.renderFailedUpload)(sarifReport, error));
                throw error;
            }
        });
    }
    getApiHelper() {
        if (!this.config.apiKey) {
            this.context.stdout.write(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.\n`);
            throw new Error('API key is missing');
        }
        return (0, api_1.apiConstructor)((0, utils_2.getBaseIntakeUrl)(), this.config.apiKey);
    }
    getMatchingSarifReports(spanTags) {
        return __awaiter(this, void 0, void 0, function* () {
            const sarifReports = (this.basePaths || []).reduce((acc, basePath) => {
                const isFile = !!upath_1.default.extname(basePath);
                if (isFile) {
                    return acc.concat(fs_1.default.existsSync(basePath) ? [basePath] : []);
                }
                return acc.concat((0, glob_1.globSync)((0, utils_1.buildPath)(basePath, '*.sarif'), { dotRelative: true }));
            }, []);
            const validUniqueFiles = [...new Set(sarifReports)].filter((sarifReport) => {
                if (this.noVerify) {
                    return true;
                }
                const validationErrorMessage = (0, validation_1.validateSarif)(sarifReport);
                if (validationErrorMessage) {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                    this.context.stdout.write((0, renderer_1.renderInvalidFile)(sarifReport, [validationErrorMessage]));
                    return false;
                }
                const potentialErrors = (0, validation_1.checkForError)(sarifReport);
                if (potentialErrors.length > 0) {
                    this.context.stdout.write((0, renderer_1.renderInvalidFile)(sarifReport, potentialErrors));
                    return false;
                }
                return true;
            });
            return validUniqueFiles.map((sarifReport) => {
                return {
                    reportPath: sarifReport,
                    spanTags,
                    service: (0, utils_2.getServiceFromSarifTool)(sarifReport),
                };
            });
        });
    }
}
exports.UploadSarifReportCommand = UploadSarifReportCommand;
UploadSarifReportCommand.paths = [['sarif', 'upload']];
UploadSarifReportCommand.usage = clipanion_1.Command.Usage({
    category: 'Static Analysis',
    description: 'Upload SARIF reports files to Datadog.',
    details: `
      This command will upload SARIF reports files to Datadog.\n
      See README for details.
    `,
    examples: [
        ['Upload all SARIF report files in current directory', 'datadog-ci sarif upload .'],
        [
            'Upload all SARIF report files in src/sarif-go-reports and src/sarif-java-reports',
            'datadog-ci sarif upload src/sarif-go-reports src/sarif-java-reports',
        ],
        [
            'Upload all SARIF report files in current directory and add extra tags globally',
            'datadog-ci sarif upload --tags key1:value1 --tags key2:value2 .',
        ],
        [
            'Upload all SARIF report files in current directory to the datadoghq.eu site',
            'DATADOG_SITE=datadoghq.eu datadog-ci sarif upload .',
        ],
    ],
});
