// @datadog/mobile-react-native/lib/commonjs/logs/eventMapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateEventMapper = exports.formatRawLogToNativeEvent = exports.formatRawLogToLogEvent = exports.formatLogEventToNativeLog = void 0;
var _EventMapper = require("../sdk/EventMappers/EventMapper");
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

const formatLogEventToNativeLog = logEvent => {
  return logEvent;
};
exports.formatLogEventToNativeLog = formatLogEventToNativeLog;
const formatRawLogToNativeEvent = rawLog => {
  return rawLog;
};
exports.formatRawLogToNativeEvent = formatRawLogToNativeEvent;
const formatRawLogToLogEvent = (rawLog, additionalInformation) => {
  return {
    ...rawLog,
    userInfo: additionalInformation.userInfo,
    attributes: additionalInformation.attributes
  };
};
exports.formatRawLogToLogEvent = formatRawLogToLogEvent;
const generateEventMapper = logEventMapper => new _EventMapper.EventMapper(logEventMapper, formatRawLogToLogEvent, formatLogEventToNativeLog, formatRawLogToNativeEvent);
exports.generateEventMapper = generateEventMapper;
