// @datadog/mobile-react-native/lib/commonjs/rum/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RumActionType = exports.PropagatorType = exports.ErrorSource = exports.ActionSource = void 0;
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */
/**
 * The entry point to use Datadog's RUM feature.
 */
/**
 * Describe the type of a RUM Action.
 */
let RumActionType = exports.RumActionType = /*#__PURE__*/function (RumActionType) {
  /** User tapped on a widget. */
  RumActionType["TAP"] = "TAP";
  /** User scrolled a view. */
  RumActionType["SCROLL"] = "SCROLL";
  /** User swiped on a view. */
  RumActionType["SWIPE"] = "SWIPE";
  /** User pressed hardware back button (Android only). */
  RumActionType["BACK"] = "BACK";
  /** A custom action. */
  RumActionType["CUSTOM"] = "CUSTOM";
  return RumActionType;
}({});
let ErrorSource = exports.ErrorSource = /*#__PURE__*/function (ErrorSource) {
  ErrorSource["NETWORK"] = "NETWORK";
  ErrorSource["SOURCE"] = "SOURCE";
  ErrorSource["CONSOLE"] = "CONSOLE";
  ErrorSource["WEBVIEW"] = "WEBVIEW";
  ErrorSource["CUSTOM"] = "CUSTOM";
  return ErrorSource;
}({});
/**
 * Type of instrumentation on the host.
 * - DATADOG: Datadog’s propagator (`x-datadog-*`)
 * - TRACECONTEXT: W3C Trace Context (`traceparent`)
 * - B3: B3 single header (`b3`)
 * - B3MULTI: B3 multiple headers (`X-B3-*`)
 */
let PropagatorType = exports.PropagatorType = /*#__PURE__*/function (PropagatorType) {
  PropagatorType["DATADOG"] = "datadog";
  PropagatorType["TRACECONTEXT"] = "tracecontext";
  PropagatorType["B3"] = "b3";
  PropagatorType["B3MULTI"] = "b3multi";
  return PropagatorType;
}({});
let ActionSource = exports.ActionSource = /*#__PURE__*/function (ActionSource) {
  ActionSource["MANUAL"] = "MANUAL";
  ActionSource["LEGACY"] = "LEGACY";
  ActionSource["BABEL"] = "BABEL";
  return ActionSource;
}({});
