// @datadog/mobile-react-native/lib/commonjs/rum/instrumentation/resourceTracking/distributedTracing/firstPartyHosts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPropagatorsForHost = exports.firstPartyHostsRegexMapBuilder = exports.NO_MATCH_REGEX = void 0;
var _InternalLog = require("../../../../InternalLog");
var _SdkVerbosity = require("../../../../SdkVerbosity");
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

// This regex does not match anything
const NO_MATCH_REGEX = exports.NO_MATCH_REGEX = new RegExp('a^');

// From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
const escapeRegExp = string => {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
};
const firstPartyHostsRegexMapBuilder = firstPartyHosts => {
  const hostsMap = {
    datadog: [],
    tracecontext: [],
    b3: [],
    b3multi: []
  };
  firstPartyHosts.forEach(host => {
    host.propagatorTypes.forEach(propagatorType => {
      hostsMap[propagatorType].push(host.match);
    });
  });
  const regexMap = [];
  Object.entries(hostsMap).forEach(([propagatorType, hosts]) => {
    if (hosts.length > 0) {
      regexMap.push({
        propagatorType: propagatorType,
        regex: firstPartyHostsRegexBuilder(hosts)
      });
    }
  });
  return regexMap;
};
exports.firstPartyHostsRegexMapBuilder = firstPartyHostsRegexMapBuilder;
const firstPartyHostsRegexBuilder = firstPartyHosts => {
  if (firstPartyHosts.length === 0) {
    return NO_MATCH_REGEX;
  }
  try {
    // A regexp for matching hosts, e.g. when `hosts` is "example.com", it will match
    // "example.com", "api.example.com", but not "foo.com".
    const firstPartyHostsRegex = new RegExp(`^(.*\\.)*(${firstPartyHosts.map(host => `${escapeRegExp(host)}$`).join('|')})`);
    firstPartyHostsRegex.test('test_the_regex_is_valid');
    return firstPartyHostsRegex;
  } catch (e) {
    _InternalLog.InternalLog.log(`Invalid first party hosts list ${JSON.stringify(firstPartyHosts)}. Regular expressions are not allowed.`, _SdkVerbosity.SdkVerbosity.WARN);
    return NO_MATCH_REGEX;
  }
};
const getPropagatorsForHost = (hostname, firstPartyHostsRegexMap) => {
  const matchedPropagatorTypes = [];
  firstPartyHostsRegexMap.forEach(({
    regex,
    propagatorType
  }) => {
    if (regex.test(hostname)) {
      matchedPropagatorTypes.push(propagatorType);
    }
  });
  return matchedPropagatorTypes.length > 0 ? matchedPropagatorTypes : null;
};
exports.getPropagatorsForHost = getPropagatorsForHost;
