// @google-cloud/run/build/src/index.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by synthtool. **
// ** https://github.com/googleapis/synthtool **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.protos = exports.WorkerPoolsClient = exports.TasksClient = exports.ServicesClient = exports.RevisionsClient = exports.JobsClient = exports.ExecutionsClient = exports.BuildsClient = exports.v2 = void 0;
const v2 = require("./v2");
exports.v2 = v2;
const BuildsClient = v2.BuildsClient;
exports.BuildsClient = BuildsClient;
const ExecutionsClient = v2.ExecutionsClient;
exports.ExecutionsClient = ExecutionsClient;
const JobsClient = v2.JobsClient;
exports.JobsClient = JobsClient;
const RevisionsClient = v2.RevisionsClient;
exports.RevisionsClient = RevisionsClient;
const ServicesClient = v2.ServicesClient;
exports.ServicesClient = ServicesClient;
const TasksClient = v2.TasksClient;
exports.TasksClient = TasksClient;
const WorkerPoolsClient = v2.WorkerPoolsClient;
exports.WorkerPoolsClient = WorkerPoolsClient;
exports.default = { v2, BuildsClient, ExecutionsClient, JobsClient, RevisionsClient, ServicesClient, TasksClient, WorkerPoolsClient };
const protos = require("../protos/protos");
exports.protos = protos;
