// @datadog/mobile-react-native/lib/commonjs/nativeModulesTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DdNativeSdkConfiguration = void 0;
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

/**
 * In this file, native modules types extend the specs for TurboModules.
 * As we cannot use enums or classes in the specs, we override methods using them here.
 */

/**
 * The entry point to use Datadog's Logs feature.
 */

/**
 * The entry point to use Datadog's Trace feature.
 */

/**
 * A configuration object to initialize Datadog's features.
 */
class DdNativeSdkConfiguration {
  constructor(clientToken, env, applicationId, nativeCrashReportEnabled, sampleRate, site, trackingConsent, additionalConfiguration // eslint-disable-next-line no-empty-function
  ) {
    this.clientToken = clientToken;
    this.env = env;
    this.applicationId = applicationId;
    this.nativeCrashReportEnabled = nativeCrashReportEnabled;
    this.sampleRate = sampleRate;
    this.site = site;
    this.trackingConsent = trackingConsent;
    this.additionalConfiguration = additionalConfiguration;
  }
}

/**
 * The entry point to initialize Datadog's features.
 */

/**
 * The entry point to use Datadog's RUM feature.
 */
exports.DdNativeSdkConfiguration = DdNativeSdkConfiguration;
